greeter/README.md

```diff
@@ -1,21 +1,706 @@
+     InvocationContext,
+     InvocationResult,
+     Provider,
+     inject,
+     ValueOrPromise,
+   } from '@loopback/context';
+   import {CachingService} from '../services/caching.service';
+   import {CACHING_SERVICE} from '../keys';
+   import {RestBindings} from '@loopback/rest';
+   ```
+
+3. Add the pre-invocation and post-invocation logics in the `intercept` function.
+
+   ```ts
+   //we need to exclude two endpoints for the interceptor:
+   const EXCLUDED_PATHS = ['/explorer/', '/explorer/openapi.json'];
+   // And then skip them in the intercept function
+   //Otherwise when hit the explorer endpoint, it's response also get cached but no timestamp in it, which breaks the cache interceptor.
+   // Add pre-invocation logic here
+   const httpReq = await invocationCtx.get(RestBindings.Http.REQUEST, {
+     optional: true,
+   });
+   /* istanbul ignore if */
+   // A workaround to skip the explorer
+   if (EXCLUDED_PATHS.includes(httpReq.path)) {
+     return next();
+   }
+   const key = httpReq.path;
+   const lang = httpReq.acceptsLanguages(['en', 'zh']) || 'en';
+   const cachingKey = `${lang}:${key}`;
+   const cachedResult = await this.cachingService.get(cachingKey);
+   if (cachedResult) {
+     console.error('Cache found for %s %j', cachingKey, cachedResult);
+     return cachedResult;
+   }
+
+   const result = await next();
+
+   // Add post-invocation logic here
+   await this.cachingService.set(cachingKey, result);
+   return result;
+   ```
+
+4. Then bind Observer and Interceptor to you app
+   in `src/application.ts` file, add these lines:
+
+   ```ts
+   this.bind(RestExplorerBindings.CONFIG).to({
+     path: '/explorer',
+   });
+   // bind interceptor and service to app
+   this.add(createBindingFromClass(CachingService, {key: CACHING_SERVICE}));
+   this.add(createBindingFromClass(CachingInterceptor));
+
+   this.component(RestExplorerComponent);
+   this.component(GreetingComponent);
+   ```
+
+### Step 4: Try it out again!
+
+Let's try it out to see the caching server and the observer that we just bind. Restart your application.
+
+#### Test cache
+
+The expire time for each result is set to 10s (10000 milliseconds by default).
+To test the cached result and its expiration, you can
+
+Try GET/ greet/{name} with a name 'loopback' and language 'en'
+You will see a result object with a timestamp field like
+
+```s
+{
+  "timestamp": "2019-10-09T20:24:00.294Z",
+  "language": "en",
+  "greeting": "Hello, loopback!"
+}
+```
+
+Try the same endpoint and same input immediately (before the cached result expires), you will see the same timestamp printed in the result, which infers that the result is NOT calculated again but from cache.
+Wait for 10s and try same endpoint again, you will see the timestamp changes, like (maybe use an image to point out the difference)
+
+```s
+{
+// please note the time here changes
+"timestamp": "2019-10-09T20:30:29.770Z",
+"language": "en",
+"greeting": "Hello, loopback!"
+}
+```
```

Comment:
I changed the code to 

```
const lang = httpReq.acceptsLanguages(['en', 'zh', 'fr']) || 'en';
```

Now I see:

```
Cache found for fr:/greet/Thor {"timestamp":"2019-10-18T17:47:08.424Z","language":"fr","greeting":"Thor, Bonjour"}
```

Seems `not too scaleable` if I have to update the interceptor every time I add a new language.

Reply:
I don't remember why the accepted languages are hardcoded in the first place, we can remove it in the workshop.
Just `const lang = httpReq.acceptsLanguages()` (you may double check if omitting the input allows all languages)

Reply:
In interceptor
I think this is easier to understand

```
 const lang = httpReq.headers['accept-language'] || 'en';
    const cachingKey = `${lang}:${key}`;
    const cachedResult = await this.cachingService.get(cachingKey);
    if (cachedResult) {
      console.error('Cache found for %s %j', cachingKey, cachedResult);
      return cachedResult;
    }

```

and in GreetingController 

```
async greet(
    @param.path.string('name') name: string,
    @param.header.string('Accept-Language') lan: string,
  ): Promise<Message> {
    
    const language: string = lan;
    const greeting = await this.greetingService.greet(language, name);

    return {
      timestamp: new Date(),
      language,
      greeting,
    };
  }
```

Reply:
@emonddr Agreed, for the workshop let's set the language as the injected header parameter.
In a real application, the language should be read from request like `httpReq.acceptsLanguages()` instead of being provided as a parameter, but switching browser's language is very problematic in the workshop, so let's change according to your plan ^ 👍